packages/openapi-v3/test/integration/controller-spec.test.ts

```diff
@@ -0,0 +1,137 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/openapi-v3
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {expect} from '@loopback/testlab';
+import {model, property} from '@loopback/repository';
+import {ParameterObject, SchemasObject} from '@loopback/openapi-v3-types';
+import {param, requestBody, getControllerSpec, post} from '../../';
+
+describe('controller spec', () => {
+  it('adds property schemas in components.schemas', () => {
+    @model()
+    class Bar {
+      @property() name: string;
+    }
+
+    @model()
+    class Baz {
+      @property() name: string;
+    }
+
+    @model()
+    class Foo {
+      @property() bar: Bar;
+      @property() baz: Baz;
+    }
+
+    class FooController {
+      @post('/foo')
+      create(
+        @requestBody({description: 'a foo instance', required: true})
+        foo: Foo,
+      ): void {}
+    }
+
+    const expectedSpec = {
+      paths: {
+        '/foo': {
+          post: {
+            responses: {},
+            requestBody: {
+              description: 'a foo instance',
+              required: true,
+              content: {
+                'application/json': {
+                  schema: {$ref: '#/components/schemas/Foo'},
+                },
+              },
+            },
+            'x-operation-name': 'create',
+          },
+        },
+      },
+      components: {
+        schemas: {
+          Bar: {title: 'Bar', properties: {name: {type: 'string'}}},
+          Baz: {title: 'Baz', properties: {name: {type: 'string'}}},
+          Foo: {
+            // guarantee `definition` is deleted
+            title: 'Foo',
+            properties: {
+              bar: {$ref: '#/components/schemas/Bar'},
+              baz: {$ref: '#/components/schemas/Baz'},
+            },
+          },
+        },
+      },
+    };
+
+    expect(getControllerSpec(FooController)).to.eql(expectedSpec);
+  });
+  it('does not produce nested definitions', () => {
+    const paramSpec: ParameterObject = {
+      name: 'foo',
+      in: 'query',
+    };
+    @model()
+    class Foo {
+      @property() bar: number;
+    }
+    @model()
+    class MyParam {
+      @property() name: string;
+      @property() foo: Foo;
+    }
+    class MyController {
+      @post('/foo')
+      foo(@param(paramSpec) body: MyParam) {}
+    }
+
+    const schemas = (getControllerSpec(MyController).components || {})
+      .schemas as SchemasObject;
```

Comment:
Uff, this is a long line doing quite a lot! Please extract it into a helper function shared by other tests and then split the implementation to multiple lines. For example:

```ts
function getControllerSchemas(controller): SchemasObject {
  const components = getControllerSpec(MyController).components;
  if (!components) return {};
  return components.schemas;
} 
```

Why is it necessary to cast `.schemas` to `SchemaObject`? I'd expect the type definitions should be telling that the result of `getControllerSpec(MyController).components.schemas` is a `SchemaObject`.